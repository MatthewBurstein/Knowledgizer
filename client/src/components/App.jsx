import React, { Component } from 'react';
import Article from './Article';

class App extends Component {
  constructor() {
    super();

    this.state = {
      articles: [{id:'', title: '', url: ''}]
    };
  }


  componentDidMount() {
    this.callApi()
      .then(res => this.setState(res))
      .catch(err => console.log(err));
  };

  async callApi() {
    const response = await fetch('/api');
    const body = await response.json();
    if (response.status !== 200) throw Error(body.message);
    return body;
  };

  render() {
    return (
      <div>
        <h2>Knowledgizer</h2>
        {this.state.articles.map(article => {
          return <Article
            key={article.id}
            title={article.title}
            url={article.url}
          />
        })}
      </div>
    )
  }
}

export default App;
