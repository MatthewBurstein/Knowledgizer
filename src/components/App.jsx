import React, { Component } from 'react';
import Article from './Article';
import ArticleContainer from './ArticleContainer';
import { delayAndRepeat } from './helpers/AppHelpers.js';

class App extends Component {
  constructor() {
    super();

    this.articles=[];
    this.state = {
      articles: []
    };
  };


  componentDidMount() {
    this.callApi()
      .then(res => {
        this.articles = this.articles.concat(res.articles);
        delayAndRepeat(this.addArticleToState, this.articles, 1000);
      })
      .catch(err => console.log(err));
  };

  addArticleToState = article => {
    this.setState(prevState => ({ articles: [...prevState.articles, article] }));
  };

  async callApi() {
    const response = await fetch('/api');
    const body = await response.json();
    if (response.status !== 200) throw Error(body.message);
    return body;
  };

  renderArticle(article) {
      return(
        <Article
          key={article.id}
          title={article.title}
          url={article.url}
          />
        )
  };

  render() {
    return (
      <div>
        <h2>Knowledgizer</h2>
        <ArticleContainer />
        {this.state.articles && this.state.articles.map(article => {
          return this.renderArticle(article)
        })}
      </div>
    )
  }
}

export default App;
