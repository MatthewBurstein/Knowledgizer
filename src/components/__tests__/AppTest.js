import React from 'react';
import App from '../App';
import { shallow, mount } from 'enzyme';

jest.useFakeTimers();

describe('App', () => {
  const articles = [
    { id:1, title: 'some title', url: 'https://en.wikipedia.org/wiki/some title'},
    { id:2, title: 'another title', url: 'https://en.wikipedia.org/wiki/another title'},
    { id: 3, title: 'a third title', url: 'https://en.wikipedia.org/wiki/a third title' }
  ];
  const app = shallow(<App />);
  app.setState({ articles });

  it('renders properly', () => {
    expect(app).toMatchSnapshot();
  });

  it('Contains an ArticleContainer component', () => {
    const articleContainer = app.find('ArticleContainer');
    expect(articleContainer).toHaveLength(1)
  })

  it('Contains an Article component for each article in `state`', () => {
    const articles = app.find('Article');
    expect(articles).toHaveLength(3);
  });

  describe('delay and repeat article rendering functionality', () => {
    let mountedApp, callApiSpy, didMountSpy, setStateSpy;
    beforeAll(() => {
      callApiSpy = jest.spyOn(App.prototype, 'callApi').mockReturnValue(Promise.resolve({articles}));
      setStateSpy = jest.spyOn(App.prototype, 'setState');
      mountedApp = mount(<App />);
    });

    it('calls call Api on mounting', () => {
      expect(callApiSpy).toHaveBeenCalledTimes(1);
    });

    it('calls setState once per second with the next article from this.articles', () => {
      expect(setStateSpy).not.toHaveBeenCalled();

      jest.runTimersToTime(1000);
      expect(setStateSpy).toHaveBeenCalledTimes(1);
      expect(mountedApp.state().articles).toContain(articles[0]);

      jest.runTimersToTime(1000);
      expect(setStateSpy).toHaveBeenCalledTimes(2);
      expect(mountedApp.state().articles).toContain(articles[1]);

      jest.runTimersToTime(1000);
      expect(setStateSpy).toHaveBeenCalledTimes(3);
      expect(mountedApp.state().articles).toContain(articles[2]);
    });
  });

});
