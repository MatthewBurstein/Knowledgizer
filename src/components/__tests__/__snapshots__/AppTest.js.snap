// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h2>
          Knowledgizer
        </h2>,
        <ArticleContainer />,
        Array [
          <Article
            title="some title"
            url="https://en.wikipedia.org/wiki/some title"
          />,
          <Article
            title="another title"
            url="https://en.wikipedia.org/wiki/another title"
          />,
          <Article
            title="a third title"
            url="https://en.wikipedia.org/wiki/a third title"
          />,
        ],
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Knowledgizer",
        },
        "ref": null,
        "rendered": "Knowledgizer",
        "type": "h2",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "class",
        "props": Object {
          "title": "some title",
          "url": "https://en.wikipedia.org/wiki/some title",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "class",
        "props": Object {
          "title": "another title",
          "url": "https://en.wikipedia.org/wiki/another title",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "3",
        "nodeType": "class",
        "props": Object {
          "title": "a third title",
          "url": "https://en.wikipedia.org/wiki/a third title",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h2>
            Knowledgizer
          </h2>,
          <ArticleContainer />,
          Array [
            <Article
              title="some title"
              url="https://en.wikipedia.org/wiki/some title"
            />,
            <Article
              title="another title"
              url="https://en.wikipedia.org/wiki/another title"
            />,
            <Article
              title="a third title"
              url="https://en.wikipedia.org/wiki/a third title"
            />,
          ],
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Knowledgizer",
          },
          "ref": null,
          "rendered": "Knowledgizer",
          "type": "h2",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "class",
          "props": Object {
            "title": "some title",
            "url": "https://en.wikipedia.org/wiki/some title",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "class",
          "props": Object {
            "title": "another title",
            "url": "https://en.wikipedia.org/wiki/another title",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "3",
          "nodeType": "class",
          "props": Object {
            "title": "a third title",
            "url": "https://en.wikipedia.org/wiki/a third title",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
